proposals/000-composite-fetcher-api.md

```diff
@@ -0,0 +1,289 @@
+  * Link extensibility can be used to add additional metadata, for example:
+    * A `ZIPFetcher` might add a `compressedLength` which could then be used by the position list factory [to address this issue](https://github.com/readium/architecture/issues/123).
+    * Something equivalent to the `Cache-Control` HTTP header could be used to customize the behavior of a parent `CachingFetcher` for a given resource.
+* (lazy) `length: Result<Long, ResourceError>`
+  * Data length from metadata if available, or calculated from reading the bytes otherwise.
+  * We must treat this value as a hint, as it might not reflect the actual bytes length.
+
+##### Methods
+
+* `read(range: Range<Long>? = null): Result<ByteArray, ResourceError>`
+  * Reads the bytes at the given `range`.
+  * When `range` is `null`, the whole content is returned.
+  * Out-of-range indexes are clamped to the available length automatically.
+* `readAsString(encoding: Encoding? = null): Result<String, ResourceError>`
+  * Reads the full content as a `String`.
+  * `encoding: Encoding? = null`
+    * Encoding used to decode the bytes.
+    * If `null`, then it is parsed from the `charset` parameter of `link.type` and falls back on UTF-8.
+* `close()`
+  * Closes any opened file handles.
+
+##### Implementations
+
+* `NullResource(link: Link)`
+  * Creates a `Resource` returning a `null` content.
+* `StringResource(link: Link, string: String)`
+  * Creates a `Resource` serving a string.
+* `BytesResource(link: Link, bytes: ByteArray)`
+  * Creates a `Resource` serving an array of bytes.
+
+
+### Examples of `Fetcher` Trees
+
+The fetcher tree created by the publication parsers can be adapted to fit the needs of different formats.
+
+#### CBZ and ZAB (Zipped Audio Book) Formats
+
+These formats are very simple, we just need to access the ZIP entries.
+
+<img src="assets/002-zip.svg">
+
+#### Audiobook Manifest
+
+The resources of a remote audiobook are accessed through HTTP requests, using an `HTTPFetcher`. However, we can implement an offline cache by wrapping the fetcher in a `CachingFetcher`.
+
+<img src="assets/002-audiobook.svg">
+
+#### LCP Protected Package (Audiobook, LCPDF, etc.)
+
+The resources of a publication protected with LCP need to be decrypted. For that, We're using a `DecryptionTransformer` through a `TransformingFetcher`. Any remote resources declared in the manifest are fetched using an `HTTPFetcher`.
+
+<img src="assets/002-lcp.svg">
+
+#### EPUB Format
+
+The EPUB fetcher is one of the most complex:
+* An `HTTPFetcher` is used for remote resources access.
+* The resources are transformed at two different levels:
+  * in the *streamer*, to decrypt the content and deobfuscate fonts,
+  * in the *navigator*, to inject the CSS and JavaScript necessary for rendering.
+
+<img src="assets/002-epub.svg">
+
+
+### Leaf Fetchers
+
+A leaf fetcher is an implementation of `Fetcher` handling the actual low-level bytes access. It doesn't delegates to any other `Fetcher`.
```

Comment:
```suggestion
A leaf fetcher is an implementation of `Fetcher` handling the actual low-level bytes access. It doesn't delegate to any other `Fetcher`.
```

Reply:
Thanks, I made some changes in my local clone so I'll do the fix there